File: phone/android/src/main/kotlin/com/digitalkey/sdk/DigitalKeySdk.kt
/**
 * DigitalKey SDK - Android SDK入口
 * 
 * 提供数字钥匙核心功能的统一入口点
 * 
 * @version 1.0.0
 * @date 2026-05-06
 */
package com.digitalkey.sdk

import android.content.Context
import com.digitalkey.sdk.key.KeyManager
import com.digitalkey.sdk.key.KeyManagerImpl
import com.digitalkey.sdk.vehicle.VehicleController
import com.digitalkey.sdk.vehicle.VehicleControllerImpl
import com.digitalkey.sdk.channel.ChannelManager
import com.digitalkey.sdk.channel.ChannelManagerImpl
import com.digitalkey.sdk.security.SecurityModule
import com.digitalkey.sdk.security.SecurityModuleImpl
import com.digitalkey.sdk.network.ApiClient
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * SDK配置参数
 * 
 * @property serverUrl 服务器地址，如 "https://api.digitalkey.cn"
 * @property appId 应用标识
 * @property clientId 客户端ID
 * @property apiKey API密钥
 * @property enableLog 是否启用日志
 */
data class SdkConfig(
    val serverUrl: String,
    val appId: String,
    val clientId: String,
    val apiKey: String,
    val enableLog: Boolean = false
)

/**
 * SDK初始化结果
 */
sealed class InitResult {
    data class Success(val sdk: DigitalKeySdk) : InitResult()
    data class Failure(val error: DigitalKeyException) : InitResult()
}

/**
 * DigitalKey SDK 主入口对象
 * 
 * 使用示例:
 * ```kotlin
 * val result = DigitalKeySdk.init(context, SdkConfig(
 *     serverUrl = "https://api.digitalkey.cn",
 *     appId = "com.example.app",
 *     clientId = "android_xxxxx",
 *     apiKey = "your_api_key"
 * ))
 * ```
 */
object DigitalKeySdk {
    
    private var _instance: DigitalKeySdkImpl? = null
    private val lock = Any()
    
    /**
     * 初始化SDK
     * 
     * @param context Android应用上下文
     * @param config SDK配置参数
     * @return 初始化结果
     */
    fun init(context: Context, config: SdkConfig): InitResult {
        synchronized(lock) {
            if (_instance != null) {
                return InitResult.Success(_instance!!)
            }
            
            try {
                val instance = DigitalKeySdkImpl(context, config)
                _instance = instance
                return InitResult.Success(instance)
            } catch (e: Exception) {
                return InitResult.Failure(
                    DigitalKeyException.Network(-1, "SDK初始化失败: ${e.message}")
                )
            }
        }
    }
    
    /**
     * 获取SDK实例
     * 
     * @return SDK实例，未初始化时返回null
     */
    fun getInstance(): DigitalKeySdkImpl? = _instance
    
    /**
     * 检查SDK是否已初始化
     */
    fun isInitialized(): Boolean = _instance != null
    
    /**
     * 销毁SDK实例
     */
    fun destroy() {
        synchronized(lock) {
            _instance?.destroy()
            _instance = null
        }
    }
}

/**
 * SDK实现类
 * 
 * 提供各模块的访问入口
 */
class DigitalKeySdkImpl(
    private val context: Context,
    private val config: SdkConfig
) {
    // API客户端
    private val apiClient: ApiClient = ApiClient(config)
    
    // 各模块实例
    private val _keyManager: KeyManager by lazy {
        KeyManagerImpl(apiClient, context)
    }
    
    private val _vehicleController: VehicleController by lazy {
        VehicleControllerImpl(apiClient, context)
    }
    
    private val _channelManager: ChannelManager by lazy {
        ChannelManagerImpl(context)
    }
    
    private val _securityModule: SecurityModule by lazy {
        SecurityModuleImpl(context)
    }
    
    /**
     * 密钥管理模块
     */
    val keyManager: KeyManager
        get() = _keyManager
    
    /**
     * 车辆控制模块
     */
    val vehicleController: VehicleController
        get() = _vehicleController
    
    /**
     * 通道管理模块
     */
    val channelManager: ChannelManager
        get() = _channelManager
    
    /**
     * 安全模块
     */
    val securityModule: SecurityModule
        get() = _securityModule
    
    /**
     * 获取SDK配置
     */
    fun getConfig(): SdkConfig = config
    
    /**
     * 销毁SDK，释放资源
     */
    fun destroy() {
        apiClient.destroy()
    }
}

/**
 * 统一异常类型
 */
sealed class DigitalKeyException : Exception() {
    /** 网络异常 */
    data class Network(val code: Int, override val message: String) : DigitalKeyException()
    
    /** 认证异常 */
    data class Auth(val code: Int, override val message: String) : DigitalKeyException()
    
    /** 密钥异常 */
    data class Key(val code: Int, override val message: String) : DigitalKeyException()
    
    /** 车辆异常 */
    data class Vehicle(val code: Int, override val message: String) : DigitalKeyException()
    
    /** 硬件异常 */
    data class Hardware(override val message: String) : DigitalKeyException()
}

/**
 * 硬件异常子类
 */
class NfcDisabledException : DigitalKeyException.Hardware("NFC未启用")
class BluetoothDisabledException : DigitalKeyException.Hardware("蓝牙未启用")
class LocationDisabledException : DigitalKeyException.Hardware("位置服务未启用")
class SecureElementException : DigitalKeyException.Hardware("安全元件不可用")
class UwbNotSupportedException : DigitalKeyException.Hardware("UWB不支持")

File: phone/ios/Sources/DigitalKeySDK/DigitalKeySDK.swift
// DigitalKeySDK.swift
// 数字钥匙SDK主入口
// 版本: 1.0.0

import Foundation
import Combine

// MARK: - SDK统一错误类型

/// 数字钥匙SDK统一错误枚举
public enum DigitalKeyError: Error, LocalizedError {
    /// 网络错误
    case network(code: Int, message: String)
    /// 认证错误
    case auth(code: Int, message: String)
    /// 密钥错误
    case key(code: Int, message: String)
    /// 车辆错误
    case vehicle(code: Int, message: String)
    /// 硬件错误
    case hardware(message: String)
    /// SDK未初始化
    case notConfigured
    /// 参数错误
    case invalidParameter(message: String)
    /// 超时
    case timeout(message: String)

    public var errorDescription: String? {
        switch self {
        case .network(let code, let msg): return "[网络错误 \(code)] \(msg)"
        case .auth(let code, let msg): return "[认证错误 \(code)] \(msg)"
        case .key(let code, let msg): return "[密钥错误 \(code)] \(msg)"
        case .vehicle(let code, let msg): return "[车辆错误 \(code)] \(msg)"
        case .hardware(let msg): return "[硬件错误] \(msg)"
        case .notConfigured: return "SDK未初始化，请先调用 DigitalKeySDK.configure()"
        case .invalidParameter(let msg): return "[参数错误] \(msg)"
        case .timeout(let msg): return "[超时] \(msg)"
        }
    }
}

// MARK: - SDK配置

/// SDK配置信息
public struct SdkConfig {
    /// 服务器地址
    public let serverUrl: String
    /// 应用标识
    public let appId: String
    /// 客户端ID
    public let clientId: String
    /// API Key
    public let apiKey: String
    /// 是否启用日志
    public let enableLog: Bool
    /// 请求超时时间（秒）
    public let timeoutInterval: TimeInterval

    public init(
        serverUrl: String,
        appId: String,
        clientId: String,
        apiKey: String,
        enableLog: Bool = false,
        timeoutInterval: TimeInterval = 30.0
    ) {
        self.serverUrl = serverUrl
        self.appId = appId
        self.clientId = clientId
        self.apiKey = apiKey
        self.enableLog = enableLog
        self.timeoutInterval = timeoutInterval
    }
}

// MARK: - SDK主类

/// 数字钥匙SDK主入口
/// 使用单例模式，通过 configure() 初始化
public class DigitalKeySDK {

    // MARK: - 单例

    /// 共享实例
    public static var shared: DigitalKeySDK {
        guard let instance = _shared else {
            fatalError("DigitalKeySDK 未配置，请先调用 DigitalKeySDK.configure()")
        }
        return instance
    }

    private static var _shared: DigitalKeySDK?

    /// SDK是否已配置
    public static var isConfigured: Bool {
        return _shared != nil
    }

    // MARK: - 属性

    /// SDK配置
    public private(set) var config: SdkConfig

    /// 密钥管理器
    public lazy var keyManager: KeyManaging = KeyManager(sdk: self)

    /// 车辆控制器
    public lazy var vehicleController: VehicleControlling = VehicleController(sdk: self)

    /// 分享管理器
    public lazy var shareManager: ShareManaging = ShareManager(sdk: self)

    /// 通道管理器
    public lazy var channelManager: ChannelManaging = ChannelManager(sdk: self)

    /// 安全模块
    public lazy var securityModule: SecurityManaging = SecurityModule(sdk: self)

    /// 网络客户端
    lazy var apiClient: ApiClient = ApiClient(config: config)

    /// 撤销信号集合
    private var cancellables = Set<AnyCancellable>()

    // MARK: - 初始化

    /// 配置SDK（仅可调用一次）
    /// - Parameters:
    ///   - serverUrl: 服务器地址，例如 "https://api.digitalkey.cn"
    ///   - appId: 应用标识，例如 "com.example.app"
    ///   - clientId: 客户端ID，例如 "ios_xxxxx"
    ///   - apiKey: API密钥
    /// - Throws: 配置失败时抛出异常
    public static func configure(
        serverUrl: String,
        appId: String,
        clientId: String,
        apiKey: String
    ) throws {
        try configure(config: SdkConfig(
            serverUrl: serverUrl,
            appId: appId,
            clientId: clientId,
            apiKey: apiKey
        ))
    }

    /// 配置SDK（完整配置）
    /// - Parameter config: SDK配置
    /// - Throws: 配置失败时抛出异常
    public static func configure(config: SdkConfig) throws {
        // 验证必填参数
        guard !config.serverUrl.isEmpty else {
            throw DigitalKeyError.invalidParameter(message: "serverUrl 不能为空")
        }
        guard !config.appId.isEmpty else {
            throw DigitalKeyError.invalidParameter(message: "appId 不能为空")
        }
        guard !config.clientId.isEmpty else {
            throw DigitalKeyError.invalidParameter(message: "clientId 不能为空")
        }
        guard !config.apiKey.isEmpty else {
            throw DigitalKeyError.invalidParameter(message: "apiKey 不能为空")
        }

        // 验证URL格式
        guard URL(string: config.serverUrl) != nil else {
            throw DigitalKeyError.invalidParameter(message: "serverUrl 格式不合法")
        }

        _shared = DigitalKeySDK(config: config)
        log("SDK配置完成: \(config.serverUrl), appId=\(config.appId)")
    }

    /// 重置SDK（用于测试或切换账号）
    public static func reset() {
        _shared = nil
        log("SDK已重置")
    }

    private init(config: SdkConfig) {
        self.config = config
    }

    // MARK: - 日志

    /// 内部日志输出
    static func log(_ message: String) {
        guard let sdk = _shared, sdk.config.enableLog else { return }
        print("[DigitalKeySDK] \(message)")
    }

    /// 内部错误日志输出
    static func logError(_ message: String) {
        guard let sdk = _shared, sdk.config.enableLog else { return }
        print("[DigitalKeySDK] ❌ \(message)")
    }
}
